### ISSUE TITLE
Loading large areas won't necessarily load the whole area

### ISSUE THREAD

**Issue by McToel** [gboeing/osmnx#791]
**Problem description**
I want to make a graph from a big part of Europe. My analysis shows, that parts of the data might be missing.

I think the issue is caused by the download module and its splitting of the area I want to download into smaller chunks.

One part of the download routine is the following function:
```python
def _make_overpass_polygon_coord_strs(polygon):
    """ ... """
    geometry_proj, crs_proj = projection.project_geometry(polygon)
    gpcs = utils_geo._consolidate_subdivide_geometry(geometry_proj)
    geometry, _ = projection.project_geometry(gpcs, crs=crs_proj, to_latlong=True)
```
I have inserted the following code in order to visualize what's happening:
```python
    import matplotlib.pyplot as plt
    import geopandas as gpd
    import cartopy.crs as ccrs
    import cartopy.feature as cfeature
    import matplotlib.pyplot as plt

    prj = ccrs.Mercator(min_latitude=40, max_latitude=60)
    ax = plt.axes(projection=prj)
    ax.add_feature(cfeature.LAND)
    ax.coastlines()
    # Manually adjust threshold for a more detailed plot
    prj.threshold = 10

    gpd.GeoSeries([polygon]).plot(ax=ax, color='gray', transform=ccrs.Geodetic(), edgecolor='blue')
    gpd.GeoSeries([geometry], crs=settings.default_crs).plot(ax=ax, edgecolor='black', facecolor='none', transform=ccrs.Geodetic())
    plt.show()
```
Which is producing the following plot:
![image](https://user-images.githubusercontent.com/36071676/148012632-4fa41bfa-0515-4dbc-97b4-2d6abbed8cd9.png)

As you might see, some chunks do not fully overlap with the original polygon. It is hard to see on this scale, so the next two plots show it a little more detailed:
![image](https://user-images.githubusercontent.com/36071676/148013356-2ba2f1d7-2f05-4a1a-9b1c-d5ec735e945d.png)
![image](https://user-images.githubusercontent.com/36071676/148013472-fdad555c-e57d-4be2-9fc1-e18d8a1704e6.png)

As far as I can tell, the same might be happening during the truncation of the graph. I guess that this issue is not noticeable with smaller (e.g. city sized) areas. For the Europe sized polygon that I use, it is also not a huge issue. The effect might be worse for other areas of the same size.

**Environment information**
python 3.9
osmnx 1.1.2
If the rest matters, I can supply it.

**Provide a complete minimal reproducible example**

I use the following code to get my graph
```python
import osmnx as ox
import shapely

polygon_wkt = 'POLYGON ((2.140369 41.378914, -3.666028 42.371166, -4.479815 48.387805, 0.320865 51.442894, 8.290385000000001 55.777294, 37.57981 55.776672, 26.074412 44.44677, 12.502788 41.900493, 2.140369 41.378914))'
polygon = shapely.wkt.loads(polygon_wkt)
graph = ox.graph_from_polygon(polygon)
```

**Comment by gboeing** [comment]
Interesting. I've played around with this a lot, but haven't noticed this issue. Can you identify its source? Is it coming in on the projection? Or the subdividing? Or the consolidation/convex-hulling? If you can trace it down to a specific bug somewhere, we can work to resolve it.

**Comment by McToel** [comment]
I think it's a result of the projection, a straight line on earth is not  a straight line in the projection. Cartopy seems to solve that by interpolating straight lines and then transforming all the points individually. I looked into the Cartopy source code and it seems to be very complex. But it works: https://scitools.org.uk/cartopy/docs/v0.15/matplotlib/intro.html#adding-data-to-the-map.

If osmnx wants to solve this problem, it would probably be wise to include cartopy as dependency. However at least on Ubuntu 20.04, Cartopy is a pain to install because it depends on a recent version of proj, which is not present in any system repositories.

**Comment by gboeing** [comment]
I'm not sure it's the result of the projection per se, because we project to UTM then reproject back to 4326 to generate the query polygons. A quick check with a comparable large, northern geography:

```python
import osmnx as ox

place = "Canada"
gdf = ox.geocode_to_gdf(place)
geom = gdf['geometry'].iloc[0]
crs = gdf.crs

geom_proj, crs_proj = ox.projection.project_geometry(geom)
geom2, _ = ox.projection.project_geometry(geom_proj, crs=crs_proj, to_crs=crs)
print((geom - geom2).area)  # 0.0

geom_proj_cs = ox.utils_geo._consolidate_subdivide_geometry(geom_proj).buffer(0)
print((geom_proj - geom_proj_cs).area)  # 3.2e-6

geom2_cs, _ = ox.projection.project_geometry(geom_proj_cs, crs=crs_proj, to_crs=crs)
print((geom - geom2_cs).area)  # 0.0002
```

Results:

  - The first test shows the the projection to/from centroid UTM zone results in a geometry diff with area = 0.
  - The second test shows that projection to UTM then running the consolidate/subdivide algorithm results in a geometry diff with area = 3.2e-6 (square meters).
  - The third test shows that the entire workflow (projection to UTM, consolidate/subdivide, project back to 4326) results in a geometry diff with area = 0.0002 ("squared decimal degrees"), which is in the ballpark of a couple hundred sq meters at the latitudes we're talking about. This is a trivial spatial error relative to the area of the original geometry, all of Canada, along the lines of 0.00001%.

Is this numerically comparable to the error magnitude you're seeing? We're probably not going to add cartopy as another dependency, and this error magnitude is probably within the acceptable tolerance, unless you're seeing a significantly larger magnitude.

Nevertheless, you can overcome any such spatial by buffering your query area slightly, downloading whatever you wanted within that geometry, then truncating it to the original desired query area.

**Comment by McToel** [comment]
I tested your code with my polygon, and it produces very different results:
```py
geom = shapely.wkt.loads('POLYGON ((2.140369 41.378914, -3.666028 42.371166, -4.479815 48.387805, 0.320865 51.442894, 8.290385000000001 55.777294, 37.57981 55.776672, 26.074412 44.44677, 12.502788 41.900493, 2.140369 41.378914))')
crs = 'EPSG:4326'

geom_proj, crs_proj = ox.projection.project_geometry(geom)
geom2, _ = ox.projection.project_geometry(geom_proj, crs=crs_proj, to_crs=crs)
print((geom - geom2).area)  # 4.568066650465099e-13

geom_proj_cs = ox.utils_geo._consolidate_subdivide_geometry(geom_proj).buffer(0)
print((geom_proj - geom_proj_cs).area)  # 0.0

geom2_cs, _ = ox.projection.project_geometry(geom_proj_cs, crs=crs_proj, to_crs=crs)
print((geom - geom2_cs).area)  # 10.497542394897875
```

I think the difference might be that in my case, the individual points that define my polygon are very far apart.

**Comment by gboeing** [comment]
> I think the difference might be that in my case, the individual points that define my polygon are very far apart.

In a quick test, it looks like this hypothesis is correct. Re-running your code snippet after interpolating 1000 vertices along your polygon eliminates the problem:

```python
import numpy as np
import osmnx as ox
import shapely

geom = shapely.wkt.loads('POLYGON ((2.140369 41.378914, -3.666028 42.371166, -4.479815 48.387805, 0.320865 51.442894, 8.290385000000001 55.777294, 37.57981 55.776672, 26.074412 44.44677, 12.502788 41.900493, 2.140369 41.378914))')
crs = 'EPSG:4326'

dists = np.linspace(0, geom.exterior.length, 1000)
geom = shapely.geometry.Polygon(geom.exterior.interpolate(d) for d in dists)

geom_proj, crs_proj = ox.projection.project_geometry(geom)
geom2, _ = ox.projection.project_geometry(geom_proj, crs=crs_proj, to_crs=crs)
print((geom - geom2).area)  # 2.5e-13

geom_proj_cs = ox.utils_geo._consolidate_subdivide_geometry(geom_proj).buffer(0)
print((geom_proj - geom_proj_cs).area)  # 2.9e-5

geom2_cs, _ = ox.projection.project_geometry(geom_proj_cs, crs=crs_proj, to_crs=crs)
print((geom - geom2_cs).area)  # 5.6e-5
```

So, the issue seems to affect very large query areas with very few vertices. The question is to what extent this is an edge case or requires an enhancement to the OSMnx codebase. I suppose the options now are:

  1) you can incorporate similar interpolation code into your script to resolve the issue
  2) we implement a similar vertex interpolation in OSMnx if the geometry exceeds some area threshold and has fewer than some vertex count threshold

**Comment by gboeing** [comment]
Actually, I should add a 3rd option to the list: OSMnx raises a warning if the geometry exceeds some area threshold and has fewer than some vertex count threshold. The warning could recommend the user interpolate more vertices, just as I demonstrated above. As this might be an edge case, this could be the best option.

**Comment by McToel** [comment]
The subdivision seems to make the problem worse
![image](https://user-images.githubusercontent.com/36071676/155766104-a855266e-6bb6-4dea-ad36-aa57d571f336.png)

This makes sense, as the subdivision is performed (right now) in latlon coordinates. The shortest path in a latlon PlateCarree does not relate to the shortest path on earth.

Solving the problem correctly would involve tons of code. In order to raise a warning, one would have to find some condition to raise it, which also seems hard to me.
I think the best solution would be to add a note to the docs of graph_from_polygon describing the issue.

**Comment by gboeing** [comment]
Would you like to open a quick PR with this doc note?

**Comment by gboeing** [comment]
Resolved by #803.
